### PR TITLE
Reorganize repository to clarify relationship betwen the devworkspace API and devfile 2.0

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR restructures the repository in order to:
- Make a clearer distinction between `DevWorkspace` and `Devfile` schemas (`DevWorkspace` is a spec for a K8S-native, API-oriented, cloud workspace specification under the form of a K8S CRD, while `Devfile 2.0` is a user-oriented format syntax describing the structure of a cloud workspace, that can be used for a wider range of tools)
- Provide `jsonpatch` rules to derive the `Devfile 2.0` schema
from the schema of the workspace structure in the `DevWorkspace` spec
- Isolate devfile-related elements (readme, jsonpatch rules and Devfile samples) in a dedicated folder,
- Generate all the Json schemas from the `DevWorkspace` API Go source code, in a single `build.sh` script
- Allow running the `build.sh` script in a docker container that already contains the prerequisites,
- Provide a `devfile 1.0.0` yaml file at the root of this repository to allow any contributor to test `DevWorkspace` or `Devfile 2.0` samples directly in [Hosted Che](che.openshift.io), using the documentation and completion features based on the generated json schemas of any PR branch.  

### Is your PR tested? Consider putting some instruction how to test your changes

Yes.

In order to test the resulting `Devfile 2.0` or `DevWorkspace` samples in a self-service Che workspace (hosted on `che.openshift.io`), just click on the button added in [this comment](https://github.com/che-incubator/devworkspace-api/pull/23#issuecomment-582954583)

As soon as the workspace is opened, you should be able to open the `yaml` files in folders `samples` or `devfile-support/samples` folders, and get `yaml` language support based on the Json schemas generated in the `schemas` folder.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+cp ${BASE_DIR}/schemas/devworkspace-template-spec.json ${BASE_DIR}/schemas/devfile.json
+
+onError() {
+  echo "Cleaning schemas/devfile.json"
+  rm -f ${BASE_DIR}/schemas/devfile.json
+}
+trap 'onError' ERR
+
+for jsonpatch in $(ls ${BASE_DIR}/devfile-support/transformation-rules/*.json 2> /dev/null)
```
Comment (amisevsk) [inline build.sh]: this logic seems strange -- why not just 
```suggestion
for jsonpatch in ${BASE_DIR}/devfile-support/transformation-rules/*.json
```

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
```
Comment (amisevsk) [inline build.sh]: This file has a lot of shellcheck errors around quoting strings. I believe it will fail if run from a path that includes a directory with a space in it.

File: devfile-support/transformation-rules/add-name-and-version.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    "op": "add",
+    "path": "/properties/schemaVersion",
+    "value": { "type": "string", "description": "Devfile schema version", "pattern": "^([2-9]+)\\.([0-9]+)\\.([0-9]+)(\\-[0-9a-z-]+(\\.[0-9a-z-]+)*)?(\\+[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?$" }
+  },
+  {
+    "op": "add",
+    "path": "/required",
+    "value": [ "name", "schemaVersion" ]
+  }
+]
```
Comment (amisevsk) [inline devfile-support/transformation-rules/add-name-and-version.json]: Trailing newline

File: devworkspace-api.code-workspace [hunk 4]
```diff
@@ -24,8 +24,8 @@
...
         "go.autocompleteUnimportedPackages": true,
         "yaml.schemas": {
-            "./generated/schema.json": "*/*devfile.yaml",
-        },
-    
+            "./schemas/devfile.json": "devfile-support/samples/*devfile.yaml",
+            "./schemas/devworkspace.json": "samples/*devworkspace.yaml",
+        }
     }
 }
```
Comment (amisevsk) [inline devworkspace-api.code-workspace]: Trailing newline

File: docker-run.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+set
```
Comment (amisevsk) [inline docker-run.sh]: Is this intentionally included?

File: docker-run.sh [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+# git ROOT directory used to mount filesystem
+GIT_ROOT_DIRECTORY=$(git rev-parse --show-toplevel)
+GO_MODULE=$(grep -e 'module ' ${GIT_ROOT_DIRECTORY}/go.mod | sed -e 's/module //')
+WORKDIR="/home/user/go/src/${GO_MODULE}"
+# Container image
+IMAGE_NAME="che-incubator/devworkspace-build-prerequisistes"
```
Comment (amisevsk) [inline docker-run.sh]: Typo? `prerequisistes -> prerequisites`

File: docker-run.sh [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+check() {
+  if [ $# -eq 0 ]; then
+    printf "%bError: %bNo script provided. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}"
+    exit 1
+  fi
+  echo "check $1"
+  if [ ! -f "$1" ]; then
+    printf "%bError: %bscript %b provided is not existing. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}" "${1}"
```
Comment (amisevsk) [inline docker-run.sh]:
```suggestion
    printf "%bError: %bscript %b provided does not exist. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}" "${1}"
```

File: docker-run.sh [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  BOLD='\033[1m'
+}
+
+check() {
+  if [ $# -eq 0 ]; then
+    printf "%bError: %bNo script provided. Command is $ docker-run.sh <script-to-run> [optional-arguments-of-script-to-run]\n" "${RED}" "${NC}"
+    exit 1
+  fi
+  echo "check $1"
+  if [ ! -f "$1" ]; then
```
Comment (amisevsk) [inline docker-run.sh]: Should also check file is executable:
```suggestion
  if [ ! -f "$1" ] || [ ! -x "$1"]; then
```

File: docker-run.sh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  WORKDIR ${WORKDIR}
+EOF
+then
+  printf "%b[OK]%b\n" "${GREEN}" "${NC}"
+  rm docker-build-log
+else
+  printf "%bFailure%b\n" "${RED}" "${NC}"
+  cat docker-build-log
+  exit 1
+fi
```
Comment (amisevsk) [inline docker-run.sh]: Indentation is off from `if` start above.

File: docker-run.sh [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  RUN apk add --no-cache --update curl bash jq go \
+  && pip3 install yq \
+  && pip3 install jsonpatch
+
+  RUN curl -JL https://github.com/operator-framework/operator-sdk/releases/download/${OPERATOR_SDK_VERSION}/operator-sdk-${OPERATOR_SDK_VERSION}-x86_64-linux-gnu -o /bin/operator-sdk && chmod a+x /bin/operator-sdk
+  RUN mkdir -p /home/user/go && chmod -R a+w /home/user
+  ENV HOME /home/user
+  ENV GOPATH /home/user/go
+  WORKDIR ${WORKDIR}
+EOF
```
Comment (amisevsk) [inline docker-run.sh]: Does building a dockerimage in this way allow caching of layers?

File: build.sh [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+BOLD='\033[1m'
+
+if ! which yq &> /dev/null
+then
+  echo
+  echo "#### ERROR ####"
+  echo "####"
+  echo "#### Please install the 'yq' tool before being able to use this script"
+  echo "#### see https://github.com/kislyuk/yq"
+  echo "#### and https://stedolan.github.io/jq/download"
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply build.sh]: this is what is installed in `docker-run.sh`:

```
  RUN apk add --no-cache --update curl bash jq go \
  && pip3 install yq \
  && pip3 install jsonpatch
```

Is it consistent with what you're saying here ?

File: docker-run.sh (hunk 6 above)
Comment (davidfestal) [reply docker-run.sh]: yes, typo. Let me fix this.

File: docker-run.sh (hunk 9 above)
Comment (davidfestal) [reply docker-run.sh]: True. Fixing.

File: docker-run.sh (hunk 10 above)
Comment (davidfestal) [reply docker-run.sh]: Yes, afaik. @benoitf Do you confirm ?

File: .theia/settings.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
+{
+  "go.lintTool": "golangci-lint",
+  "go.lintFlags": "--fast",
+  "yaml.schemas": {
+      "./schemas/devfile.json": "devfile-support/samples/*devfile.yaml",
+      "./schemas/devworkspace.json": "samples/*devworkspace.yaml"
+  }
+}
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply .theia/settings.json]: fixed

File: docker-run.sh (hunk 6 above)
Comment (davidfestal) [reply docker-run.sh]: Fixed

File: docker-run.sh (hunk 9 above)
Comment (davidfestal) [reply docker-run.sh]: Fixed

File: build.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
```
Comment (benoitf) [inline build.sh]: you may add copyright headers

File: build.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply build.sh]: Yes, the stedolan version of yq is written in c and has different syntax from what we're used to (`jq`-like). `kislyuk/yq` is a python wrapper around jq and works how we intend.

File: docker-run.sh [hunk 14]
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+#
+# Copyright (c) 2019 Red Hat, Inc.
```
Comment (benoitf) [inline docker-run.sh]:
```suggestion
# Copyright (c) 2020 Red Hat, Inc.
```

File: build.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply build.sh]: Ahh I misread. Disregard this message -- I somehow assumed the second line was referring to http://mikefarah.github.io/yq/

[the remark this answers is not in the document]
Comment (benoitf) [reply build.sh]: @amisevsk I think you're talking about two impls of `yq` but here links are one for `jq` and one for `yq`
no ?

File: docker-run.sh (hunk 5 above)
Comment (benoitf) [reply docker-run.sh]: when running the script I've tons of output so I would say it's more for debugging purpose and hasn't been removed ?

File: docker-run.sh (hunk 10 above)
Comment (davidfestal) [reply docker-run.sh]: At least from what I experience locally it's not rebuilding the image if I run the build script a second time.

File: build.sh (hunk 13 above)
Comment (davidfestal) [reply build.sh]: done

File: docker-run.sh (hunk 10 above)
Comment (benoitf) [reply docker-run.sh]: yes

```bash
docker build -t foo -<<EOF
FROM python:3-alpine
RUN apk add --no-cache --update bash jq go
RUN apk add --no-cache --update curl
EOF
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM python:3-alpine
 ---> a1cd5654cf3c
Step 2/3 : RUN apk add --no-cache --update bash jq go
 ---> Running in 008183dd2fc5
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/15) Installing bash (5.0.11-r1)
Executing bash-5.0.11-r1.post-install
(2/15) Installing libgcc (9.2.0-r3)
(3/15) Installing libstdc++ (9.2.0-r3)
(4/15) Installing binutils (2.33.1-r0)
(5/15) Installing gmp (6.1.2-r1)
(6/15) Installing isl (0.18-r0)
(7/15) Installing libgomp (9.2.0-r3)
(8/15) Installing libatomic (9.2.0-r3)
(9/15) Installing mpfr4 (4.0.2-r1)
(10/15) Installing mpc1 (1.1.0-r1)
(11/15) Installing gcc (9.2.0-r3)
(12/15) Installing musl-dev (1.1.24-r0)
(13/15) Installing go (1.13.4-r1)
(14/15) Installing oniguruma (6.9.4-r0)
(15/15) Installing jq (1.6-r0)
Executing busybox-1.31.1-r9.trigger
OK: 428 MiB in 50 packages
Removing intermediate container 008183dd2fc5
 ---> a0553fff2d8d
Step 3/3 : RUN apk add --no-cache --update curl
 ---> Running in 41fa96d1fb49
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/3) Installing nghttp2-libs (1.40.0-r0)
(2/3) Installing libcurl (7.67.0-r0)
(3/3) Installing curl (7.67.0-r0)
Executing busybox-1.31.1-r9.trigger
OK: 429 MiB in 53 packages
Removing intermediate container 41fa96d1fb49
 ---> 725ca1c61c87
Successfully built 725ca1c61c87
Successfully tagged foo:latest





docker build -t foo -<<EOF
FROM python:3-alpine
RUN apk add --no-cache --update bash jq go
RUN apk add --no-cache --update curl
EOF
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM python:3-alpine
 ---> a1cd5654cf3c
Step 2/3 : RUN apk add --no-cache --update bash jq go
 ---> Using cache
 ---> a0553fff2d8d
Step 3/3 : RUN apk add --no-cache --update curl
 ---> Using cache
 ---> 725ca1c61c87
Successfully built 725ca1c61c87
Successfully tagged foo:latest
```

**Comment by benoitf** [comment]
would be nice to include in the README the instructions to build

even if it's just
```bash
./docker-run.sh build.sh
```

File: build.sh (hunk 1 above)
Comment (davidfestal) [reply build.sh]: The suggested change triggers the following error when no `.json` file is found in the `transformation-rules` folder:
```
usage: jsonpatch [-h] [--indent INDENT] [-b] [-i] [-v] ORIGINAL [PATCH]
jsonpatch: error: argument PATCH: can't open '/home/user/go/src/github.com/che-incubator/devworkspace-api/devfile-support/transformation-rules/*.json': [Errno 2] No such file or directory: '/home/user/go/src/github.com/che-incubator/devworkspace-api/devfile-support/transformation-rules/*.json'
```

File: docker-run.sh (hunk 5 above)
Comment (davidfestal) [reply docker-run.sh]: Exact, just a rest from `set -x`

File: build.sh (hunk 1 above)
Comment (amisevsk) [reply build.sh]: In that case I would go for something like
```bash
patches_dir="${BASE_DIR}/devfile-support/transformation-rules/"
readarray -d '' patches < <(find "$patches_dir" -name '*.json' -print0)
for jsonpatch in "${patches[@]}"; do
  #...
done
```
The current approach fails locally for me since I have an alias defined for `ls`.

File: build.sh (hunk 2 above)
Comment (davidfestal) [reply build.sh]: I fixed most of those. Only warnings are remaining, about no-used variables related to colors at the beginning.

**Comment by davidfestal** [comment]
> would be nice to include in the README the instructions to build
> 
> even if it's just
> 
> ```shell
> ./docker-run.sh build.sh
> ```

done

**Comment by davidfestal** [comment]
@amisevsk is it OK for you now.

@l0rd Are you OK to merge this ?

File: README.md [hunk 15]
```diff
@@ -1,2 +1,30 @@
-# devworkspace-api
-K8S Api for a cloud develoment workspace specification [Draft proposal]
+# DevWorkspace API
+
+K8S-native Api for a cloud develoment workspace specification [Draft proposal]
```
Comment (l0rd) [inline README.md]: This sentence already in the repo description. I think it's redundant here. And I would rephrase it "Kube-native API for cloud development workspaces specification".

File: README.md [hunk 16]
```diff
@@ -1,2 +1,30 @@
...
+
+Sources for this API are defined in Go code, starting from the
+[devworkspace_types.go source file](pkg/apis/workspaces/v1alpha1/devworkspace_types.go)
+
+From these Go sources, several files are generated:
+- a K8S Custom Resource definition with an embedded OpenApi schema,
```
Comment (l0rd) [inline README.md]:
```suggestion
- A Kubernetes Custom Resource Definition with an embedded OpenApi schema,
```

File: README.md [hunk 17]
```diff
@@ -1,2 +1,30 @@
...
+- json schemas (in the [schemas](schemas) folder) derived from the above custom resource definition, to specify the synax of:
```
Comment (l0rd) [inline README.md]:
```suggestion
- json schemas (in the [schemas](schemas) folder) generated from the above Custom Resource Definition, to specify the syntax of:
```

File: README.md [hunk 18]
```diff
@@ -1,2 +1,30 @@
...
+  - the DevWorkspace CRD itself
+  - the format for templates of a DevWorkspace (template content of a workspace, without runtime information),
```
Comment (l0rd) [inline README.md]:
```suggestion
  - the DevWorkspaceTemplate CRD (a workspace content, without runtime information),
```

File: README.md [hunk 19]
```diff
@@ -1,2 +1,30 @@
...
+  - the Devfile 2.0 format, which is expected to be derived from the `DevWorkspace` API.
```
Comment (l0rd) [inline README.md]:
```suggestion
  - the Devfile 2.0.0 format, which is generated from the `DevWorkspace` API.
```

File: README.md [hunk 20]
```diff
@@ -1,2 +1,30 @@
...
+
+Generated files are created by a build script (see section [How to build](#how-to-build)).
+
+### Devfile 2.0 support
```
Comment (l0rd) [inline README.md]:
```suggestion
### Devfile 2.0.0 support
```

File: README.md [hunk 21]
```diff
@@ -1,2 +1,30 @@
...
+
+A Subset of this `DevWorkspace` API defines a structure (workspace template content), which is also at the core of the **Devfile 2.0** format specification.
+For more information about this, please look into
+the [Devfile support Readme](devfile-support/README.md)
+
+## How to build
+
+In order to build the CRD and the various schemas, you don't need to install any pre-requisite apart from `docker`).
```
Comment (l0rd) [inline README.md]:
```suggestion
In order to build the CRD and the various schemas, you don't need to install any pre-requisite apart from `docker`.
```

File: docker-run.sh [hunk 22]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
```
Comment (l0rd) [inline docker-run.sh]: Instead of requiring users to build the docker image we should have a `Dockerfile.build` file and just publish the image on quay.io. That would have the merit to:
- avoid building the image to build the image :-)
- `docker run ...` would be the build command, and that's is easier to understand and customise (for example replacing `docker` with `buildah`)

Comment (l0rd) [reply docker-run.sh]: But that can be done in another PR

Comment (davidfestal) [reply docker-run.sh]: I agree 100 % with that, knowing that this image would also be usable in a `dockerImage` component of the repo `devfile.yaml` to allow testers to run the build and the test their changes themselves.

But I didn't want to wait for this before submitting this PR.
